s3_storage_provider.py

```diff
@@ -244,19 +268,59 @@ def _stream_to_producer(reactor, producer, body, status=None, timeout=None):
             after being paused
     """
 
+    if not status:
+        status = _ProducerStatus()
+
+    try:
+        if s3._cse_client:
+            stream_body_with_cse(body, producer, reactor, s3, timeout, status)
+        else:
+            stream_body(body, producer, reactor, status, timeout)
+    except Exception:
+        reactor.callFromThread(producer._error, Failure())
+    finally:
+        reactor.callFromThread(producer._finish)
+        if body:
+            body.close()
+
+def stream_body(body, producer, reactor, status, timeout):
     # Set when we should be producing, cleared when we are paused
     wakeup_event = producer.wakeup_event
-
     # Set if we should stop producing forever
     stop_event = producer.stop_event
 
     if not status:
         status = _ProducerStatus()
```

Comment:
I think we'll always have a status passed in now, we should probably add docstrings to `stream_body` and `stream_body_with_cse`.